app/V1Module/presenters/ExercisesPresenter.php

```diff
@@ -657,5 +675,86 @@ public function actionDetachPipeline(string $id, string $pipelineId) {
     $this->sendSuccessResponse($this->exerciseViewFactory->getExercise($exercise));
   }
 
+  // **********************************************
+  // ******************** TAGS ********************
+  // **********************************************
+
+  public function checkAllTags() {
+    if (!$this->exerciseAcl->canViewAllTags()) {
+      throw new ForbiddenRequestException("You are not allowed to view all tags");
+    }
+  }
+
+  /**
+   * Get list of global exercise tag names which are currently registered.
+   * @GET
+   */
+  public function actionAllTags() {
+    $tags = $this->exerciseTags->findAllDistinctNames();
+    $this->sendSuccessResponse($tags);
+  }
+
+  public function checkAddTag(string $id) {
+    $exercise = $this->exercises->findOrThrow($id);
+    if (!$this->exerciseAcl->canAddTag($exercise)) {
+      throw new ForbiddenRequestException("You are not allowed to add tag to the exercise");
+    }
+  }
+
+  /**
+   * Add tag with given name to the exercise.
+   * @POST
+   * @param string $id
+   * @param string $name
+   * @throws BadRequestException
+   * @throws NotFoundException
+   * @throws ForbiddenRequestException
+   * @throws InvalidArgumentException
+   */
+  public function actionAddTag(string $id, string $name) {
+    $exercise = $this->exercises->findOrThrow($id);
+    $tag = $this->exerciseTags->findByNameAndExercise($name, $exercise);
+    if ($tag !== null) {
+      throw new BadRequestException("Tag already exists on exercise");
+    }
+
+    if (empty($name)) {
+      throw new BadRequestException("Tag cannot be empty");
+    }
+
+    if (strlen($name) > 32) {
+      throw new InvalidArgumentException("name", "tag name too long (exceeds 32 characters)");
```

Comment:
This can be done using the `@Param` annotation.

Reply:
You are absolutely right